blocks/blockstore/bloom_cache.go

```diff
@@ -0,0 +1,175 @@
+package blockstore
+
+import (
+	"github.com/ipfs/go-ipfs/blocks"
+	key "github.com/ipfs/go-ipfs/blocks/key"
+	lru "gx/ipfs/QmVYxfoJQiZijTgPNHCHgHELvQpbsJNTg6Crmc3dQkj3yy/golang-lru"
+	bloom "gx/ipfs/QmWQ2SJisXwcCLsUXLwYCKSfyExXjFRW2WbBH5sqCUnwX5/bbloom"
+	context "gx/ipfs/QmZy2y8t9zQH2a1b8q2ZSLKp17ATuJoCNxxyMFG5qFExpt/go-net/context"
+	ds "gx/ipfs/QmfQzVugPq1w5shWRcLWSeiHF4a2meBX7yVD8Vw7GWJM9o/go-datastore"
+)
+
+// BloomCached returns Blockstore that caches Has requests using Bloom filter
+// Size is size of bloom filter in bytes
+func BloomCached(bs Blockstore, bloomSize, lruSize int) (*bloomcache, error) {
+	bl, err := bloom.New(float64(bloomSize), float64(7))
+	if err != nil {
+		return nil, err
+	}
+	arc, err := lru.NewARC(lruSize)
+	if err != nil {
+		return nil, err
+	}
+	bc := &bloomcache{blockstore: bs, bloom: bl, arc: arc}
+	bc.Invalidate()
+	go bc.Rebuild()
+
+	return bc, nil
+}
+
+type bloomcache struct {
+	bloom  *bloom.Bloom
+	active bool
+
+	arc *lru.ARCCache
+	// This chan is only used for testing to wait for bloom to enable
+	rebuildChan chan struct{}
+	blockstore  Blockstore
+
+	// Statistics
+	hits   uint64
+	misses uint64
+}
+
+func (b *bloomcache) Invalidate() {
+	b.rebuildChan = make(chan struct{})
+	b.active = false
+}
+
+func (b *bloomcache) BloomActive() bool {
+	return b.active
+}
+
+func (b *bloomcache) Rebuild() {
+	ctx := context.TODO()
+	evt := log.EventBegin(ctx, "bloomcache.Rebuild")
+	defer evt.Done()
+
+	ch, err := b.blockstore.AllKeysChan(ctx)
+	if err != nil {
+		log.Errorf("AllKeysChan failed in bloomcache rebuild with: %v", err)
+		return
+	}
+	for key := range ch {
+		b.bloom.AddTS([]byte(key)) // Use binary key, the more compact the better
+	}
+	close(b.rebuildChan)
+	b.active = true
+}
+
+func (b *bloomcache) DeleteBlock(k key.Key) error {
+	if has, ok := b.hasCached(k); ok && !has {
+		return ErrNotFound
+	}
+
+	b.arc.Remove(k) // Invalidate cache before deleting.
+	err := b.blockstore.DeleteBlock(k)
+	if err == nil {
+		b.arc.Add(k, false)
+	} else if err == ds.ErrNotFound || err == ErrNotFound {
+		b.arc.Add(k, false)
+		return ErrNotFound
+	}
+	return err
+}
+
+// if ok == false has is inconclusive
+// if ok == true then has respons to question: is it contained
+func (b *bloomcache) hasCached(k key.Key) (has bool, ok bool) {
+	if k == "" {
+		return true, true
+	}
+	if b.active {
```

Comment:
Does it? I would prefer not to introduce locking here as it is quite slow when compared with speed of bloom.

It would only introduce race if bloom was being deactivated (it isn't in yet, but that doesn't matter) and the pointer to `bloom` was being changed. If we don't change the pointer and we acquire internal bloom's RWLock, we can then work on the internals of the bloom.

I will leave a comment about that if it seems good to you.